### PR TITLE
fix: When Installing Calendar App, Set Primary Calendar As Selected Calendar

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR sets the primary/default calendar of a calendar integration as the first selected calendar on install. 

The calendars that have this ability are:
- Google Calendar
- Outlook
- Zoho
- Lark

Once #12011 is merged we can add a test for this

CleanShot 2023-11-08 at 22 17 13

Fixes # (issue)

## Requirement/Documentation

- If there is a requirement document, please, share it here.
- If there is ab UI/UX design document, please, share it here.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?

- Install the calendars that are included in this PR
- On the installed page, the primary calendar should already be selected

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/app-store/office365calendar/api/callback.ts [hunk 1]
```diff
@@ -74,6 +76,50 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
     },
   });
 
+  // Set the isDefaultCalendar as selectedCalendar
+  // If a user has multiple calendars, keep on making calls until we find the default calendar
+  let defaultCalendar: OfficeCalendar | undefined = undefined;
+  let requestUrl = "https://graph.microsoft.com/v1.0/me/calendars?$select=id,isDefaultCalendar";
+  let finishedParsingCalendars = false;
+
+  while (!finishedParsingCalendars) {
```
Comment (joeauyeung) [inline packages/app-store/office365calendar/api/callback.ts:85]: MS Graph only returns 10 calendars at a time. If for any reason the primary calendar is not a part of the first call then continue making calendar calls.

**Comment by supalarry** [comment]
Tested locally - can confirm that after installing Google Calendar and going to "Installed apps" the toggle for my e-mail for Google Calendar was on.

File: packages/app-store/office365calendar/api/callback.ts [hunk 2]
```diff
@@ -74,6 +76,50 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
+  if (defaultCalendar?.id && req.session?.user?.id) {
+    await prisma.selectedCalendar.create({
+      data: {
+        userId: req.session?.user.id,
+        integration: "office365_calendar",
+        externalId: defaultCalendar.id,
+        credentialId: credential.id,
+      },
+    });
+  }
```
Comment (exception) [inline packages/app-store/office365calendar/api/callback.ts:121]: `suggestion-non-blocking`: I'd prefer encapsulating this logic into its own function, and only running it if the actual `req.session.user.id` exists, if not running it is useless no?

```ts
const findAndStoreDefaultCalendar = async (accessToken: string, credentialId: string, userId: string) => {
  let requestUrl = "https://graph.microsoft.com/v1.0/me/calendars?$select=id,isDefaultCalendar";
  
  try {
    while (true) {
      const response = await fetch(requestUrl, {
        method: "GET",
        headers: {
          Authorization: `Bearer ${accessToken}`,
          "Content-Type": "application/json",
        },
      });

      const { value: calendars, "@odata.nextLink": nextLink } = await handleErrorsJson<{ value: OfficeCalendar[]; "@odata.nextLink"?: string }>(response);
      const defaultCalendar = calendars.find(calendar => calendar.isDefaultCalendar);

      if (defaultCalendar) {
        await prisma.selectedCalendar.create({
          data: {
            userId,
            integration: "office365_calendar",
            externalId: defaultCalendar.id,
            credentialId,
          },
        });

        break;
      }

      if (!nextLink) {
        break;
      }

      requestUrl = nextLink;
    }
  } catch (error) {
    throw error;
  }
}

if (req.session?.user?.id) {
  await findAndStoreDefaultCalendar(responseBody.access_token, credential.id, req.session.user.id);
}
```

File: packages/app-store/zohocalendar/api/callback.ts [hunk 3]
```diff
@@ -64,7 +64,36 @@ async function getHandler(req: NextApiRequest, res: NextApiResponse) {
...
+  const calendarResponse = await fetch("https://calendar.zoho.com/api/v1/calendars", {
+    method: "GET",
+    headers: {
+      Authorization: `Bearer ${key.access_token}`,
+      "Content-Type": "application/json",
+    },
+  });
+  const data = await calendarResponse.json();
+
+  const primaryCalendar = data.calendars.find((calendar: any) => calendar.isdefault);
```
Comment (exception) [inline packages/app-store/zohocalendar/api/callback.ts:85]: `nitpick`: Please ensure a type, even if `{ isdefault: boolean }`

Comment (emrysal) [inline packages/app-store/zohocalendar/api/callback.ts:85]: Valid eslint warning, non-blocking tho.

**Review by emrysal — APPROVED** [review]

File: packages/app-store/larkcalendar/api/callback.ts [hunk 4]
```diff
@@ -89,6 +89,32 @@ async function getHandler(req: NextApiRequest, res: NextApiResponse) {
...
+        await prisma.selectedCalendar.create({
+          data: {
+            userId: req.session?.user.id,
+            integration: "lark_calendar",
+            externalId: primaryCalendar.data.calendars.calendar.calendar_id as string,
+            credentialId: currentCredential?.id,
+          },
+        });
+      }
+    }
```
Comment (exception) [inline packages/app-store/larkcalendar/api/callback.ts:116]: `suggestion`: Please check for `req.session.user.id` first, no need to run the logic if we're not gonna store it.

**Review by exception — CHANGES_REQUESTED** [review]

File: packages/app-store/zohocalendar/api/callback.ts (hunk 3 above)
Comment (emrysal) [reply packages/app-store/zohocalendar/api/callback.ts:85]: Uncertain types aren't a good idea, `any` is actually better as it conveys something is really "unknown" - I did mark this as a nitpick as this is a potential issue.

The right fix for an issue like this is to parse the JSON response and validate the isdefault key is there using zod, then you don't need to type any - then throw a descriptive error when this fails the parsing.

File: packages/app-store/larkcalendar/api/callback.ts (hunk 4 above)
Comment (emrysal) [reply packages/app-store/larkcalendar/api/callback.ts:116]: See L106 - the extra ? isn't necessary here though @joeauyeung, req.session.user.id is sure to be defined here.

File: packages/app-store/office365calendar/api/callback.ts (hunk 2 above)
Comment (emrysal) [reply packages/app-store/office365calendar/api/callback.ts:121]: Even better, looking at the entire code it looks like the credential create will also fail if there is no session user, so the ideal fix is to return early out of the entire callback when the session user id isn't truthy.

**Pull request merged**